contracts/contracts/src/functioncall/gpactv2/CallPathCallExecutionTreeV2.sol

```diff
@@ -0,0 +1,170 @@
+    uint256 private constant HASH_SIZE = 32;
+
+    /**
+     * Based on a call path, extract the function call hash of a function to be executed.
+     *
+     * @param _callExecutionTree Data blob representing a call tree.
+     * @param _callPath Path of the function hash to be extracted.
+     * @return functionCallHash Function call hash of the function represented by the _callPath.
+     */
+    function extractTargetFromCallGraph(
+        bytes memory _callExecutionTree,
+        uint256[] memory _callPath
+    ) internal pure returns (bytes32 functionCallHash) {
+        uint8 callTreeType = BytesUtil.bytesToUint8(
+            _callExecutionTree,
+            OFFSET_OF_TYPE
+        );
+        if (callTreeType == SINGLE_LAYER) {
+            return
+                extractTargetFromCallGraphSingleLayer(
+                    _callExecutionTree,
+                    _callPath
+                );
+        }
+        if (callTreeType == MULTI_LAYER) {
+            extractTargetFromCallGraphMultiLayer(_callExecutionTree, _callPath);
+        }
+        revert("Unknown call tree type");
+    }
+
+    /**
+     * Extract the function call hash assuming a call tree with root and called functions only,
+     * in the Simple Encoding format.
+     *
+     * @param _callExecutionTree Data blob representing a call tree.
+     * @param _callPath Path of the function hash to be extracted.
+     * @return functionCallHash Function call hash of the function represented by the _callPath.
+     */
+    function extractTargetFromCallGraphSingleLayer(
+        bytes memory _callExecutionTree,
+        uint256[] memory _callPath
+    ) internal pure returns (bytes32 functionCallHash) {
+        require(
+            _callPath.length == 1,
+            "Call Path Error: Multiple layers in a single layer call tree"
+        );
+        uint256 calledFunc = _callPath[0];
+
+        uint8 numFuncsCalled = BytesUtil.bytesToUint8(
+            _callExecutionTree,
+            OFFSET_NUM_FUNCS_CALLED
+        );
+        require(
+            calledFunc <= numFuncsCalled,
+            "Call Path Error: calling beyond end of call tree"
+        );
+        uint256 index = OFFSET_FUNCS_CALLED_ARRAY + calledFunc * HASH_SIZE;
+        functionCallHash = BytesUtil.bytesToBytes32(_callExecutionTree, index);
+    }
+
+    /**
+     * Extract the function call hash assuming a call tree with root and called functions only,
+     * in the Arbitrary Call Tree Encoding format.
+     *
+     * @param _callExecutionTree Data blob representing a call tree.
+     * @param _callPath Path of the function hash to be extracted.
+     * @return functionCallHash Function call hash of the function represented by the _callPath.
+     */
+    function extractTargetFromCallGraphMultiLayer(
+        bytes memory _callExecutionTree,
+        uint256[] memory _callPath
+    ) internal pure returns (bytes32 functionCallHash) {
+        uint256 index = 1;
```

Comment:
can I confirm this `index` is meant to be 1 and not 0 (unlike [here in CallPathCallExecutionTree](https://github.com/ConsenSys/gpact/blob/main/contracts/contracts/src/functioncall/gpact/CallPathCallExecutionTree.sol#L39))? is there an assumption in this function that should be asserted through a `require` to ensure this condition? similar to line 86